### PR TITLE
[core] Corrupted resource stays in the cache until expiration

### PR DESCRIPTION
**Pull request by tmpsantos**
Any resource that makes to the cache is only evicted when it expires or the cache runs out of space. In case of critical resources like the style or tile source, this can potentially "brick" the map.

We should evict corrupted resources as we find them.

### CODE REVIEW

**Comment by tmpsantos** [comment]
Work in progress, will finish the implementation and write tests.

File: platform/default/mbgl/storage/offline_database.cpp [hunk 1]
```diff
@@ -184,6 +184,15 @@ std::pair<bool, uint64_t> OfflineDatabase::putInternal(const Resource& resource,
     return { inserted, size };
 }
 
+void OfflineDatabase::remove(const Resource& resource) {
+    // clang-format off
+    Statement stmt = getStatement("DELETE FROM resources WHERE url = ?1");
```
Comment (jfirebaugh) [inline platform/default/mbgl/storage/offline_database.cpp]: Should handle tiles too.

**Comment by tmpsantos** [comment]
:eyes: @jfirebaugh 

I'm not particularly happy with 137cdfd4fa2f6f, but did not find a better way since unlike the raster tiles, we parse vector tiles on a worker.

**Comment by jfirebaugh** [comment]
I'm not convinced that this adds enough value to justify the additional complexity, especially given that the added code path is infrequently exercised and therefore has a relatively high chance of regressions that worsen a failure case than recover from it. I'm imagining something like a bug in the `reportBad` pathway that turns what would have been "map doesn't render / is missing tiles" into "app crashes".

It wouldn't have saved us in the case of #5722 -- the crash there happened before `reportBad` would have been called. And even in the case that it does result in successfully evicting a bad resource from the cache, what prevents the bad resource simply being re-downloaded from the server and re-cached, ad infinitum?

**Comment by tmpsantos** [comment]
@jfirebaugh I do agree that this code won't be frequently exercised but I believe we have enough unit tests for it and we can probably add more if needed.

Currently without this patch Mapbox GL Native will crash if it fails to decompress something from the cache. This is very unlikely but nevertheless a possible scenario.

Also on the unlikely event that we get a tile source we cannot parse, that will pretty much brick the map until it expires.

The current approach is very conservative and will simply purge the item from the cache and won't cause a retry. If we are talking a about a mobile app here, it will only unbrick the map if you set the style again (probably will require you to close/open the app).

The follow-up from this will be report "resource loading failed" as a signal to the client, so it can decide to reload the map. Currently we only do that when failing to load the style.

**Comment by mb12** [comment]
@tmpsantos and @jfirebaugh several private wifi networks (esp on airports and coffee shops) always return an http sign in page for every https/http request until one signs in. Since the offline download does not do any validation,  downloads initiated under such network conditions would result in this code path being used quite often.

**Comment by 1ec5** [comment]
> The follow-up from this will be report "resource loading failed" as a signal to the client, so it can decide to reload the map.

#1804

> several private wifi networks (esp on airports and coffee shops) always return an http sign in page for every https/http request until one signs in.

Previous discussion about captive networks: #943.

**Comment by tmpsantos** [comment]
:eyes: @kkaefer

File: test/storage/offline_database.cpp [hunk 2]
```diff
@@ -224,15 +224,38 @@ TEST(OfflineDatabase, PutResourceNoContent) {
...
     response.noContent = true;
 
-    db.put(resource, response);
-    auto res = db.get(resource);
-    EXPECT_EQ(nullptr, res->error);
-    EXPECT_TRUE(res->noContent);
-    EXPECT_FALSE(res->data.get());
+    Resource style { Resource::Style, "http://example.com/" };
+    db.put(style, response);
+    EXPECT_FALSE(db.get(style));
```
Comment (kkaefer) [inline test/storage/offline_database.cpp]: Could you please add a note about why `EXPECT_FALSE` is correct here. This is not obvious from just reading this code.

**Comment by kkaefer** [comment]
Instead of tacking on a `std::function` to every Response object, we should funnel the notification that a resource is bad directly to the filesource.

I think the captive portal scenario that @mb12 describes is valid, and we should add tests for this. Instead of providing a callback that allows reporting bad data, we could also introduce very simple sanity checks directly in the `OnlineFileSource` (e.g. make sure JSON files start with a valid JSON character, images have a magic header, and PBF files are parseable. Of course, this means that we could still get invalid data delivered from our server, but as @jfirebaugh says, in the case of #5722, this wouldn't have helped us.

**Comment by jfirebaugh** [comment]
> Instead of tacking on a std::function to every Response object, we should funnel the notification that a resource is bad directly to the filesource.

I'd be more amenable to this feature if we did that. [Code like this](https://github.com/mapbox/mapbox-gl-native/blob/416cdab5460cb2eee1a99483f958d044c3a11874/platform/default/default_file_source.cpp#L88-L90) makes me think "how do we know the `this` bound in the lamba is still valid when it's called"?

**Comment by tmpsantos** [comment]
@jfirebaugh @kkaefer thanks for reviewing. With the new Actor model we can do much better here, gonna rewrite it.

**Comment by kkaefer** [comment]
I still believe that this is a valid issue, in particular bad captive portals poisoning the cache. If we implement some way of deleting bad resources from the cache, we're also going to trigger a re-request. This means we'll have to make sure we don't enter a loop that in quick succession requests + deletes the same item over and over again. An exponential back off (like we have for failing requests) could work here.

**Comment by tmpsantos** [comment]
Abandoned.

**Pull request closed without merging**